File: algorithms/2_medium/N/01358_number_of_substrings_containing_all_three_characters/solution.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        last_seen = [-1] * 3
        cnt = 0

        for (i, ch) in enumerate(s):
            last_seen[ord(ch) - ord('a')] = i
            cnt += 1 + min(last_seen)

        return cnt

File: algorithms/2_medium/N/01358_number_of_substrings_containing_all_three_characters/solution.rs
impl Solution {
    pub fn number_of_substrings(s: String) -> i32 {
        assert!(s.len() < i32::MAX as usize);
        let mut last_seen = [-1_i32; 3];
        let mut cnt = 0_i32;

        for (i, &byte) in s.as_bytes().into_iter().enumerate() {
            assert!(byte >= b'a' && byte <= b'c');
            last_seen[(byte - b'a') as usize] = i as i32;
            cnt += 1 + *last_seen.iter().min().unwrap();
        }

        cnt
    }
}
